### ISSUE TITLE
linter: support for --type-check to ignore `ignorePatterns` and be based on tsconfig.json

### ISSUE THREAD

**Issue by oriSomething** [oxc-project/oxc#16748]
The reason, in our (where I work) project we have some generated ts files that we don't want to lint, but do need to type check. This is not the only case. But type checking is important for all files that are referenced from `tsconfig.json`.
I'm aware it might introduce a bizarre API. But for us, linting from "oxlint" (or "eslint") isn't 1:1 as type checking.

[reference](https://x.com/oriSomething/status/1998285329657364627)

**Comment by camc314** [comment]
Thanks for suggesting this!

I think the following should work:

1. in your code generation tool, add `/* oxlint-disable */` at the top of each of your generated files. This will disable all lint rules for those files whie still allowing type checking to run.
2. Please checkout https://github.com/camc314/oxlint-repros/tree/main/issue-16748 for an example

Let me know if this solve the issue

**Comment by oriSomething** [comment]
I need the opposite. **Not** skipping type check while skipping linting

**Comment by camc314** [comment]
Is that not what my suggestion does?

We get type errors from `generated/*` but not lint errors on those files

Without disable comment:

```
$ pnpm run lint

> issue-16748@1.0.0 lint /Users/cameron/github/camc314/oxlint-repros/issue-16748
> oxlint --type-check --type-aware


  × eslint(no-unused-vars): Variable 'generatedValue' is declared but never used. Unused variables should start with a '_'.
    ╭─[generated/types.ts:21:7]
 20 │ // A lint error will also be reported here, since `generatedValue` is unused
 21 │ const generatedValue: string = true;
    ·       ───────┬──────
    ·              ╰── 'generatedValue' is declared here
    ╰────
  help: Consider removing this declaration.

  × eslint(no-unused-vars): Variable 'userId' is declared but never used. Unused variables should start with a '_'.
    ╭─[src/index.ts:18:7]
 17 │ // Type error: assigning string to number
 18 │ const userId: number = "not-a-number";
    ·       ───┬──
    ·          ╰── 'userId' is declared here
    ╰────
  help: Consider removing this declaration.

  × typescript(TS2322): Type 'boolean' is not assignable to type 'string'.
    ╭─[generated/types.ts:21:7]
 20 │ // A lint error will also be reported here, since `generatedValue` is unused
 21 │ const generatedValue: string = true;
    ·       ──────────────
    ╰────

  × typescript(TS2322): Type 'string' is not assignable to type 'number'.
    ╭─[src/index.ts:18:7]
 17 │ // Type error: assigning string to number
 18 │ const userId: number = "not-a-number";
    ·       ──────
    ╰────

Found 0 warnings and 4 errors.
Finished in 252ms on 2 files using 12 threads.
 ELIFECYCLE  Command failed with exit code 1.
```

With disable comment:

```
$ pnpm run lint

> issue-16748@1.0.0 lint /Users/cameron/github/camc314/oxlint-repros/issue-16748
> oxlint --type-check --type-aware


  × eslint(no-unused-vars): Variable 'userId' is declared but never used. Unused variables should start with a '_'.
    ╭─[src/index.ts:18:7]
 17 │ // Type error: assigning string to number
 18 │ const userId: number = "not-a-number";
    ·       ───┬──
    ·          ╰── 'userId' is declared here
    ╰────
  help: Consider removing this declaration.

  × typescript(TS2322): Type 'boolean' is not assignable to type 'string'.
    ╭─[generated/types.ts:22:7]
 21 │ // A lint error will also be reported here, since `generatedValue` is unused
 22 │ const generatedValue: string = true;
    ·       ──────────────
    ╰────

  × typescript(TS2322): Type 'string' is not assignable to type 'number'.
    ╭─[src/index.ts:18:7]
 17 │ // Type error: assigning string to number
 18 │ const userId: number = "not-a-number";
    ·       ──────
    ╰────

Found 0 warnings and 3 errors.
Finished in 168ms on 2 files using 12 threads.
 ELIFECYCLE  Command failed with exit code 1.
```

In the second lint run, I've added an `/*oxlint-disable*/` to the top of the file, and the lint error about `generatedValue` being an unused value disappears.

**Comment by oriSomething** [comment]
Does this type check also files in `ignorePatterns`?

**Comment by camc314** [comment]
> Does this type check also files in `ignorePatterns`?

No

**Comment by oriSomething** [comment]
This is the main issue :/

**Comment by camc314** [comment]
Adding support for something like this is going to be really confusing for users. They're going to need to look at their gitignore, .oxlintrc.json, tsconfigs all to work out what files are linted and what files are ignored. This adds also adds a complex maintence burden to keep all the different ways users could want this to work up to date. For example, what if someone has a file excluded from their tsconfig, but included in linting? do we skip linting this file?

Since the requirements here are:
1. want to lint+typecheck a set of files
2. want to skip linting a subset of these files.

My suggestion is you remove the file from `ignorePatterns`, and instead use `/* oxlint-disable */` comments at the top of generated the file.

This way, the file is typechecked, but not linted, which should achieve the result you're looking for.

**Comment by oriSomething** [comment]
I'm closing the issue because I understand it won't be resolved. But it means specific for us "type checking" isn't apply to us. Thank you for your time 🙏

**Comment by camc314** [comment]
I'm happy to consider this request, but I need to understand why ignore patterns and disable comments wont work first?

I've created a repository with an example of how to use these two together? does that approach not work for you?

**Comment by oriSomething** [comment]
Since we're using some code generated TypeScript files. Not all tools are written by us, so we don't have a control over that. But I can say for sure, sometimes because of misconfiguration we would get type errors. I'm sorry, I can't give much more details than this

**Comment by oriSomething** [comment]
I'll try to check other ways to achieve this. It's not high priority, but we do want to try using this in CI

**Comment by camc314** [comment]
> Not all tools are written by us, so we don't have a control over that.

Ah I see, so you don't have access to change the tool that's doing the code generation?

It might be worth just running a script across all the generated files to add `oxlint-disable` at the start?

```js
#!/usr/bin/env node
const fs = require('fs');
const path = require('path');

const TARGET_EXTENSIONS = ['.ts', '.js'];
const HEADER = '/* oxlint-disable */\n';

// Include `.d.ts` explicitly
function isTargetFile(file) {
  if (file.endsWith('.d.ts')) return true;
  return TARGET_EXTENSIONS.includes(path.extname(file));
}

function processFile(filePath) {
  let content = fs.readFileSync(filePath, 'utf8');

  // Skip if header already exists
  if (content.startsWith(HEADER.trim())) return;

  fs.writeFileSync(filePath, HEADER + content, 'utf8');
  console.log('Updated:', filePath);
}

function walk(dir) {
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const fullPath = path.join(dir, entry.name);

    if (entry.isDirectory()) {
      walk(fullPath);
    } else if (entry.isFile() && isTargetFile(entry.name)) {
      processFile(fullPath);
    }
  }
}

// Allow path from CLI args
const rootDir = process.argv[2] || process.cwd();

if (!fs.existsSync(rootDir)) {
  console.error('Directory does not exist:', rootDir);
  process.exit(1);
}

console.log('Scanning:', rootDir);
walk(rootDir);
console.log('Done.');
```

Hopefully you can understand my concern about trying to support all the different combinations of ignorePatterns, tsconfig excludes and includes, gitignores, includes ect.

**Comment by oriSomething** [comment]
It has a drawback, as it can mess up watchers, and then it means I needed another temporary folder for that. I will check for a different options. Thanks anyway

**Comment by camc314** [comment]
:+1: I see, mind explaining the watchers problem? The more info I have about this sort of problem, the easier it is to make a design decision/prioritize implementing this or an alternative solution

**Comment by oriSomething** [comment]
TSC watchers are broken when same files changes in a very short time

**Comment by icopp** [comment]
I've been trying to use `--type-check` with a Next.js setup, and run into this problem because Next has a bunch of generated files that are and have to be gitignored because they churn constantly (`next-env.d.ts`, `.next/types`) but break type-checking if they aren't included in Typescript builds. It makes `--type-check` unusable if you're using Next+Typescript functionality. The script suggestion above doesn't help because the Next dev server regenerates these files on the fly when it compiles changes.
